nightly.swift

```diff
@@ -33,24 +45,50 @@ extension URL {
         return self
     }
     
-    func followingRedirects() -> URL? {
+    func followingRedirects() -> RedirectResult {
         let semaphore = DispatchSemaphore(value: 0)
         
         let follower = RedirectFollower()
         let session = URLSession(configuration: .default, delegate: follower, delegateQueue: nil)
+        var result: RedirectResult?
         
-        let task = session.dataTask(with: self) { (_, response, error) in
+        let task = session.dataTask(with: self) { (data, response, error) in
+            
+            if let error = error {
+                result = .unknownError(error)
+            }
+            
+            if let lastURL = follower.lastURL, self.removingGitExtension().absoluteString != lastURL.absoluteString {
+                result = .redirected(lastURL)
+            }
+            
+            let httpResponse = response as? HTTPURLResponse
+            
+            if let statusCode = httpResponse?.statusCode {
+                switch statusCode {
+                case 404:
+                    result = .notFound
+                    
+                case 429:
```

Comment:
I remember something about GitHub and rate-limiting when we were discussing it in the server project. I don't think GitHub actually sends a 429 but instead sends a 403 error with some extra headers indicating that a rate limit error has occurred.

See [this page](https://developer.github.com/v3/#rate-limiting). Search for "If you exceed the rate limit" for the relevant piece. You can also find reference to this in the Server project [here](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/24d69ba52019c7c11e502c2433ff301e2c2d118a/Sources/App/Core/Github.swift#L61), and tested [here](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/24d69ba52019c7c11e502c2433ff301e2c2d118a/Tests/AppTests/GithubTests.swift#L199) and [here](https://github.com/SwiftPackageIndex/SwiftPackageIndex-Server/blob/24d69ba52019c7c11e502c2433ff301e2c2d118a/Tests/AppTests/GithubTests.swift#L213).

Reply:
I had found those exact same docs but as per one of our build logs found [here](https://github.com/SwiftPackageIndex/PackageList/runs/895191442?check_suite_focus=true) you can see for the actual webpage it *does* come through as a 429. It does appear that this code is working as expected!

Reply:
To be clear, this appear to be a difference between an API rate limit and the "silent" webpage rate limit. Documentation doesn't apply to the latter as it's not expected that any normal user would hit it.

Reply:
Then we may need to take another look at the Server project! 😬 Although, from the look of it it doesn't seem like we're actually checking status codes at all, and just checking for the header.

When we implemented that, it was definitely returning a 403 as I remember it tripped us up, so the 429 change may be recent.

Reply:
It could be a recent change, but I'm not so sure -- I did check on my machine and when making requests to the website you get no rate limit headers either. I genuinely think it's just a difference in behaviour between official APIs and this unofficial scraping.

I'm happy that the code here is working though

Reply:
I'm pretty sure the difference is between api and web server calls.